Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/SqlDialect.cs

```diff
@@ -50,5 +53,67 @@ public virtual string DelimitIdentifier (string identifier)
       return "[" + identifier + "]";
     }
 
+    public IDbDataParameter CreateDataParameter (IDbCommand command, IStorageTypeInformation storageTypeInformation, string parameterName, object? value)
+    {
+      ArgumentUtility.CheckNotNull("command", command);
+      ArgumentUtility.CheckNotNull("storageTypeInformation", storageTypeInformation);
+      ArgumentUtility.CheckNotNullOrEmpty("parameterName", parameterName);
+
+      var convertedValue = storageTypeInformation.ConvertToStorageType(value);
+
+      var parameter = command.CreateParameter();
+      parameter.ParameterName = GetParameterName(parameterName);
+      parameter.Value = convertedValue;
+      parameter.DbType = storageTypeInformation.StorageDbType;
+      if (storageTypeInformation.StorageTypeLength.HasValue)
+      {
+        var parameterSize = storageTypeInformation.StorageTypeLength.Value;
+        if (parameterSize == SqlStorageTypeInformationProvider.StorageTypeLengthRepresentingMax)
+        {
+          var fulltextCompatibleMaxLength = storageTypeInformation.StorageDbType switch
+          {
+              DbType.AnsiString => 8000,
+              DbType.String => 4000,
+              DbType.AnsiStringFixedLength  => default(int?), // represents char(...), which do not support "max"
+              DbType.StringFixedLength => default(int?), // represents nchar(...), which do not support "max"
+              _ => default(int?)
+          };
+
+          if (fulltextCompatibleMaxLength.HasValue)
+          {
+            var hasFullTextCompatibleMaxLength = convertedValue switch
+            {
+                DBNull => true,
+                null => true,
+                string stringValue => stringValue.Length <= fulltextCompatibleMaxLength.Value,
+                char[] charsValue => charsValue.Length <= fulltextCompatibleMaxLength.Value,
+                _ => throw new NotSupportedException($"The type '{convertedValue.GetType()}' is not fulltext compatible.")
```

Comment:
```suggestion
                _ => false
```
We should keep `false` instead of exception because not all string-cases would be checked for string-ness and it is weird to throw if the value type is e.g. int and the DbType is AnsyString but not if the value type is int and the DbType is AnsiStringFixedLength.

Reply:
We only enter this code block if there is a `fulltextCompatibleMaxLength`, which is only the case for `DbTypes` `String` and `AnsiString`. If the converted value is neither `string` nor `char[]` nor `null`/`DbNull`, the error is that the (.NET) storage type is not compatible with the `DbType`, which is only relevant here because we worry about fulltext search.
If there is a crazy `IStorageTypeInformation` that claims its `StorageDbType` is `DbType.String`, but it converts values to `byte[]`, this will trigger the exception, regardless of the actual length of the `byte[]`.

Reply:
I understand that it's likely a bug if this happens. My intention here is that the place guarding against this isn't well defined here and we already have a simple and well-defined behavior: if it's a string and the storagetype length is below the max-size, use the storage type length and otherwise leave the original default. If we add a guard here, some error scenarios would fail here and others, which don't trigger max length but have incompatible types, would fail elsewhere. that would be inconsistent and make it more difficult to identy the root cause of the issue.

The second part was the message itself which implies that the value must be fulltext compatible while it's only about the parameter length. 

So, we the general idea being that we want to only inject special handling for some very specific usecases also means we shouldn't touch other usecases but rather pass them through unchanged. 

I can't say how much of this is due to the fact that that was the original behavior and how much is a general practice. It's very well possible that a competely greenfield design might result in different concepts, but those would probably work with an entirely differnt code structure, making the datatype / length / fulltext-option very specific. Since we don't do this here, it's best to see the maxlength for the fulltext compatible parameter as an add-on for text and nothing else.

Some more background information on the orginal requirements can be found here https://re-motion.atlassian.net/issues/RM-3211

Reply:
I reverted to `false`, and also renamed the variable to `replaceMaxSizeWithFulltextCompatibleSize`, so that false is the semantically correct result for the condition in question.